### PR TITLE
Add UnorderedMapInsertOps

### PR DESCRIPTION
**Pull request by e10harvey**
@masterleinad, @fnrizzi: This never got merged into PR200 via https://github.com/marcinwrobel1986/kokkos-core-wiki/pull/1.

### CODE REVIEW

File: docs/source/API/containers/Unordered-Map.rst [hunk 1]
```diff
@@ -64,8 +64,10 @@ Description
    .. cppkokkos:kokkosinlinefunction:: UnorderedMapInsertResult insert(key) const;
 
       Insert the given key into the map with a default constructed value
+      Optionally specify the operator, op, used for combining values if
```
Comment (fnrizzi) [inline docs/source/API/containers/Unordered-Map.rst:67]: i don't see `op` in this overload?

File: docs/source/API/containers/Unordered-Map.rst [hunk 2]
```diff
@@ -127,6 +145,40 @@ There are 3 potential states for every insertion which are reported by the ``Uno
...
+    noop_type noop;
+    parallel_for(N, KOKKOS_LAMBDA (uint32_t i) {
+      map.insert(i, values(i), noop);
+    });
+    // OR;
+    parallel_for(N, KOKKOS_LAMBDA (uint32_t i) {
+      map.insert(i, values(i));
+    });
+  
+Insertion using UnorderedMapInsertOpTypes::AtomicAdd
```
Comment (fnrizzi) [inline docs/source/API/containers/Unordered-Map.rst:162]: UnorderedMapInsertOpTypes::AtomicAdd -> ``UnorderedMapInsertOpTypes::AtomicAdd``

File: docs/source/API/containers/Unordered-Map.rst [hunk 3]
```diff
@@ -127,6 +145,40 @@ There are 3 potential states for every insertion which are reported by the ``Uno
...
+-------------------------------------------------------
+
+The behavior from ``Insertion using the default UnorderedMapInsertOpTypes::NoOp`` hold true with the
```
Comment (fnrizzi) [inline docs/source/API/containers/Unordered-Map.rst:165]: i would make this a link to the corresponding subsection

File: docs/source/API/containers/Unordered-Map.rst [hunk 4]
```diff
@@ -114,8 +116,24 @@ Description
...
----------
+.. cppkokkos:struct:: template <class ValueTypeView, class ValuesIdxType> UnorderedMapInsertOpTypes
+
+   :tparam ValueTypeView: The UnorderedMap value array type.
+
+   :tparam ValuesIdxType: The index type for lookups in the value array.
+
+   .. rubric:: Public Methods
+
+   The first key inserted stores the associated value (default op).
```
Comment (fnrizzi) [inline docs/source/API/containers/Unordered-Map.rst:127]: does this sentence belong here or should go somewhere else?

File: docs/source/API/containers/Unordered-Map.rst [hunk 5]
```diff
@@ -114,8 +116,24 @@ Description
...
 
-Insertion
----------
+.. cppkokkos:struct:: template <class ValueTypeView, class ValuesIdxType> UnorderedMapInsertOpTypes
+
+   :tparam ValueTypeView: The UnorderedMap value array type.
+
+   :tparam ValuesIdxType: The index type for lookups in the value array.
+
+   .. rubric:: Public Methods
```
Comment (fnrizzi) [inline docs/source/API/containers/Unordered-Map.rst:125]: why methods if below we are describing nested structs?

File: docs/source/API/containers/Unordered-Map.rst (hunk 4 above)
Comment (fnrizzi) [reply docs/source/API/containers/Unordered-Map.rst:128]: I dont quite get why this: The first key inserted stores the associated value (default op). is here?

Comment (e10harvey) [reply docs/source/API/containers/Unordered-Map.rst:128]: Yes. That's the default behavior of the unordered map, the first k,v pair to 'win' the insertion is what sticks. All subsequent insertions will be discarded. How can we word this more clearly?

Comment (e10harvey) [reply docs/source/API/containers/Unordered-Map.rst:128]: Actually, this isn't new behavior. Let me check for existing wording.

Comment (e10harvey) [reply docs/source/API/containers/Unordered-Map.rst:128]: We currently state: `Insert the given key/value pair into the map`

File: docs/source/API/containers/Unordered-Map.rst [hunk 6]
```diff
@@ -114,8 +115,27 @@ Description
...
+
+   :tparam ValueTypeView: The UnorderedMap value array type.
+
+   :tparam ValuesIdxType: The index type for lookups in the value array.
+
+   .. rubric:: *Public* Insertion Operator Types
+
+   .. cppkokkos:struct:: NoOp
+
+   The first key inserted stores the associated value (default op).
```
Comment (nmm0) [inline docs/source/API/containers/Unordered-Map.rst:128]: This needs to be tabbed over to be part of the struct documentation

File: docs/source/API/containers/Unordered-Map.rst [hunk 7]
```diff
@@ -114,8 +115,27 @@ Description
...
+
+   .. cppkokkos:struct:: AtomicAdd
+
+   Duplicate key insertions sum values together.
```
Comment (nmm0) [inline docs/source/API/containers/Unordered-Map.rst:132]: Same here

**Review by nmm0 — CHANGES_REQUESTED** [review]
Can you rebase?

**Review by ajpowelsnl — APPROVED** [review]

**Review by nmm0 — APPROVED** [review]
:+1:

**Pull request merged**